Fixes #14565 - Empty DocType folders hidden

xrasod opened the pull request:
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes [Issue #14565](https://github.com/umbraco/Umbraco-CMS/issues/14565)

### Description
~~I have added an additional method on the contentTypeService combines the check for any contentTypes with a check for any entityContainers. This is probably not the most performant way of doing this as this results in two checks. But during my quick overview I could not find any existing check for this. Please feel free to suggest an edit if I'm mistaken here..~~


ContentTypeTreeController now also checks for the existence of any folders when setting the hasChildren property on the tree. Previously this would only check for any contentTypes. 


~~I've added tests to the contentTypeService tests-class which pass on my local machine.~~

To verify fix, 

1. Fresh install of umbraco. 
2. Under settings, 'Document Types' should have no arrow, because it's empty. 
3. Create a folder under Document Types, the folder should be visible and 'Document Types' should have an expand-arrow. 
4. Reload the browser window. 'Documen Types' should have an arrow and should have the created folder as a descendant. 

Reproduction steps for this bug can be found in the original issue.

xrasod commented:
Strange build errors? Seems unrelated to anything included in this PR to my eyes. Perhaps a rerun would resolve it?

mikecp commented:
Hi @xrasod ,

Thanks a lot for this fix! Someone from the team will have a look at it soon.
Don't worry too much about the build errors indeed, sometimes it might be that the runner had some hickups 😅

CHeers!

xrasod commented:
Right, ok. Took a closer look at the build error, and it is indeed related to the changes here. I've updated the interface for contenttypeservice, and I'm guessing this is pulling on a trip wire to prevent unwanted changes. 

Perhaps I'm missing a required step for this action, or it just needs to be manually handled. 
I've discussed this change with myself, and I did of course contemplate just inlining this check in the tree controller. This would likely avoid tripping the API compat trip wire and this would all be fine and dandy. A point could definitely be made for not introducing a method on the service which is only used once.

My initial reasoning here is that I felt that this could be a worthwhile check to integrate into the service, even if this is the only current use case for it. It also enabled me to write unit tests for this.

I also introduced some inconsistencies with the readlock, which would have to be adjusted. 

However, I have instead decided to revert the service changes and inline the check for now. 
This means there are no more unit tests for this use case, and I will instead update the steps to verify the fix in the PR.

JasonElkin commented:
Thanks @xrasod for the fix, and for the clear testing/reproduction steps too!

This works great!
